pkgs/build-support/kernel/closure.rs

```diff
@@ -0,0 +1,184 @@
+        }
+    }
+
+    fn mod_path(&self, module : &OsString) -> Option<ModPath> {
+        let fname = self.modinfo(module,"filename")?;
+        let fname = OsStr::from_bytes(&fname);
+        Some(
+            if fname == OsString::from("(builtin)") {
+                ModPath::Builtin
+            } else {
+                let path = Path::new(&fname);
+                ModPath::Path(path.to_path_buf())
+            }
+        )
+    }
+
+    fn copy_mod(&self, module : &OsString){
+        if let Some(modpath) = self.mod_path(&module) {
+            match modpath {
+                ModPath::Path(source) => {
+                    let target = self.out_dir.join(
+                        source.strip_prefix(&self.kernel).unwrap()
+                        );
+                    let target_dir = target.parent().unwrap();
+                    fs::create_dir_all(&target_dir).unwrap();
+                    fs::copy(&source,&target).unwrap();
+                }
+                ModPath::Builtin => { },
+            }
+        }
+    }
+
+    fn copy_firmware(&self, module : &OsString) {
+        if let Some(out) = self.modinfo(&module,"firmware") {
+            let firmware : Vec<OsString> = out
+                .split(|c| *c == b'\n')
+                .filter(|bs| !bs.starts_with(b"name:"))
+                .map(|bs| OsString::from_vec(Vec::from(bs)) )
+                .collect();
+            let source_dir = self.firmware_dir.join("lib/firmware");
+            let target_dir = self.out_dir.join("lib/firmware");
+            for fw in firmware {
+                let target = target_dir.join(&fw);
+                fs::create_dir_all(&target).unwrap();
+                let res = fs::copy(source_dir.join(&fw),&target);
+                if let Err(err) = res {
+                    match err.kind() {
+                        ErrorKind::NotFound => {
+                            eprint!("WARNING: missing firmware ");
+                            eprint_os(&fw);
+                            eprint!("\nfor module ");
+                            eprint_os(&module);
+                        }
+                        _ => exit(1)
+                    }
+                }
+            }
+        }
+    }
+
+    fn mod_resolve(&self, module : &OsString) -> Option<OsString> {
+        self.modinfo(&module,"name").map(|bs| OsString::from_vec(bs))
+    }
+
+    fn mod_dep(&self, module : &OsString)
+        -> HashSet<OsString> {
+
+        let dep_list = self.modinfo(module,"depends").unwrap();
+        dep_list
+            .split(|c| *c == b',')
+            .map(|bs|
+                 OsString::from_vec(
+                    bs.iter().cloned()
+                    .filter(|c| *c != b'\n')
+                    .collect()
+                    )
+                 )
+            .filter(|s| !s.is_empty())
+            .collect()
+    }
+
+}
+
+
+fn eprint_os<S : AsRef<OsStr>>(s : S) {
+    std::io::stderr().write_all(s.as_ref().as_bytes()).unwrap();
```

Comment:
probably don’t need `as_ref()` because rust allows all functions on the reference type transitively.

Reply:
The explicit as_ref was needed because S would otherwise require an OsStringExt constraint, but this function is gone now anyway